geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderAndCQDurableClientDUnitTest.java

```diff
@@ -0,0 +1,470 @@
+  }
+
+  public static boolean isTempQueueUsed() {
+    return ParallelGatewaySenderAndCQDurableClientDUnitTest.IS_TEMP_QUEUE_USED;
+  }
+
+  MemberVM getServerToStop(List<MemberVM> list) {
+    for (MemberVM member : list) {
+      if (!member.invoke(this::getPrimaryBucketList).isEmpty()) {
+        return member;
+      }
+    }
+    return null;
+  }
+
+  private void startDurableClient()
+      throws Exception {
+    int locatorPort = locatorSite2.getPort();
+    clientSite3SubscriptionQueue = clusterStartupRule.startClientVM(7, ccf -> ccf
+        .withPoolSubscription(true).withLocatorConnection(locatorPort).withCacheSetup(c -> c
+            .set("durable-client-id", DURABLE_CLIENT_ID)));
+  }
+
+  private void createDurableCQs(String... queries) {
+    clientSite3SubscriptionQueue.invoke(() -> {
+      assertThat(ClusterStartupRule.getClientCache()).isNotNull();
+      QueryService queryService = ClusterStartupRule.getClientCache().getQueryService();
+      CqAttributesFactory cqAttributesFactory = new CqAttributesFactory();
+      ParallelGatewaySenderAndCQDurableClientDUnitTest.cqListener =
+          new CqListenerTestReceivedEvents();
+      cqAttributesFactory.addCqListener(cqListener);
+
+      for (String query : queries) {
+        CqQuery cq = queryService.newCq(query, cqAttributesFactory.create(), true);
+        cq.execute();
+      }
+      ClusterStartupRule.getClientCache().readyForEvents();
+    });
+  }
+
+  private Set<Integer> getPrimaryBucketList() {
+    PartitionedRegion region = (PartitionedRegion) Objects
+        .requireNonNull(ClusterStartupRule.getCache()).getRegion("test1");
+    return new TreeSet<>(region.getDataStore().getAllLocalPrimaryBucketIds());
+  }
+
+  /**
+   * Check that all events are received on durable client that registered cq's
+   */
+  private void checkCqEvents(int expectedNumberOfEvents) {
+    // Check if number of events is correct
+    clientSite3SubscriptionQueue.invoke(() -> await().untilAsserted(() -> assertThat(
+        ParallelGatewaySenderAndCQDurableClientDUnitTest.cqListener.getNumEvents())
+            .isEqualTo(expectedNumberOfEvents)));
+  }
+
+  /**
+   * Checks that the bucketToTempQueueMap for a partitioned region
+   * that holds events for buckets that are not available locally, is empty.
+   */
+  public static void validateBucketToTempQueueMap(String senderId, boolean shouldBeEmpty) {
+    final int finalSize = sizeOfBucketToTempQueueMap(senderId);
+    if (shouldBeEmpty) {
+      assertEquals("Expected elements in TempQueueMap: " + 0
+          + " but actual size: " + finalSize, 0, finalSize);
+    } else {
+      assertThat(finalSize).isNotEqualTo(0);
+    }
+  }
+
+  public static int sizeOfBucketToTempQueueMap(String senderId) {
+    GatewaySender sender = getGatewaySender(senderId);
+    int size = 0;
+    Set<RegionQueue> queues = ((AbstractGatewaySender) sender).getQueues();
+    for (Object queue : queues) {
```

Comment:
Making this `for (RegionQueue queue : queues) {` the cast on line 290 can be removed.